src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi

```diff
@@ -78,17 +77,21 @@ cdef class _AioCall:
         """Destroys the corresponding Core object for this RPC."""
         grpc_call_unref(self._grpc_call_wrapper.call)
 
-    cdef AioRpcStatus _cancel_and_create_status(self, object cancellation_future):
-        """Cancels the RPC in Core, and return the final RPC status."""
-        cdef AioRpcStatus status
+    def cancel(self, AioRpcStatus status):
```

Comment:
100% agree

Reply:
Im wondering if it makes sense to pass the whole `status`, because it will be yes or yes  - if Im not missing something - filled during the happy path here [1].

So maybe we would need to just give as a parameters the status and the details.

[1] https://github.com/grpc/grpc/pull/21506/files#diff-d07d1d5928e4bc4defeb2e1218c57186R152

Reply:
Maybe I'm too naive on that, but my idea would be that once we would start canceling explicitly the RPC calls by using the cancel method provided by the `AioCall` would be Core who would tell you the status of finished RPC.

With this, you would be reducing the code paths where the status is being managed, so having only one place the code executed after awaiting the Core.

Having the code path for managing the status reduced to only one place, the `AioCall.cancel` method would have more sense of accepting a strict list of parameters related only to cancellation attributes.

Reply:
There are three cancellation behaviors we need to handle:
1. Cancellation from Core;
2. Task cancelled;
3. Injected `CancelledError` from parent coroutine (see [test case](https://github.com/grpc/grpc/pull/21506/files#diff-ebe6c75bf1347f8032019b99fbf4a8bdR149-R170)).

Based on the comment, your proposal is that we convert behavior 2 into behavior 1. But that will make us vulnerable to behavior 3, which is not a rare pattern. On the other hand, if we use the try-catch clause and explicit conversion, I'm certain we should cover all the cases.

The cancellation mechanism I implemented in last streaming API PR is a complex version of your proposal (due to cancellation_future). It is stable and working until I found it doesn't handle behavior 3.

Reply:
Yes, I'm suggesting implementing the use case 2 as 1, but IMO the important thing is what to do with use case 3, do we need to mask the `asyncio.CanceledError` as a `RpcError` I'm inclined to believe that better no, maybe Im missing something. So in this assert [1], I would expect to catch an `asyncio.CancelledError`.

By raising a generic `RpcError` we are opening the box for having the following two problems:

- Local coroutine within the task reacting upon the `RpcError` as it was a normal `RpcError` while the parent task asked explicitly for canceling it. 
- Having the parent task the necessity of understanding what `RpcError` semantically means.

Implementing the use case 2 as 1 is an implementation detail, as long as we are able to detect that is a local cancellation so converting the exception to a `RpcError` I would say that is not gonna be a big deal for me. 

[1] https://github.com/grpc/grpc/pull/21506/files#diff-ebe6c75bf1347f8032019b99fbf4a8bdR168

Reply:
The only drawback of not masking the `asyncio.CancelledError` as a simple `RpcError` is a likely impact on the cancellation method of the interceptors, which is basically based on the idea that a simple cancellation of a task would be enough for retrieving a `RpcError`.

By making the change that I'm proposing won't be an easy way for doing so and at the same time guarantee that `asyncio.CancelledError` are not masked as `RpcError`

Reply:
Regarding this objection, that is the unique one, I would advocate for having some feedback from @gnossen if he is ok with the current design, go ahead!

Reply:
It make sense that masking `CancelledError` as `RpcError` might work against the user's mind. If a user is calling `cancel` method explicitly, I believe they probably don't want to retry. In that case, `CancelledError` does bring value by early exiting the RPC. I would like to explore this idea a bit, that we convert all local cancellation `RpcError` to `CancelledError`.

About cancellation from parent coroutine, we might not able to differentiate a call cancellation and a parent coroutine cancellation. So, if retry is needed, the retry logic will apply to both cases or none of them.

```Python
async def coro_1(stub, request):
    call = stub.UnaryCall(request)

    async def cancel_soon():
        await asyncio.sleep(1)  # Or other business logic
        call.cancel()
    
    asyncio.create_task(cancel_soon())

    await call
    # 1. Are we expecting `CancelledError` or `RpcError`?
    # 2. Is the `CancelledError` generate by `cancel()` or parent?
    # 3. Should this coroutine end silently?

async def coro_2(stub, request):
    task = asyncio.create_task(coro_1(stub, request))
    await asyncio.wait_for(task, 1)
```

> Having the parent task the necessity of understanding what RpcError semantically means.

Cancellation is not the only trigger for `RpcError` exception. Application needs to be aware of it.

Reply:
> It make sense that masking CancelledError as RpcError might work against the user's mind. If a user is calling cancel method explicitly, I believe they probably don't want to retry. In that case, CancelledError does bring value by early exiting the RPC. I would like to explore this idea a bit, that we convert all local cancellation RpcError to CancelledError

I'm starting to believe that this is the way to go

> About cancellation from parent coroutine, we might not able to differentiate a call cancellation and a parent coroutine cancellation. So, if retry is needed, the retry logic will apply to both cases or none of them.

I would say that most of the retries would come because of a server cancelation, not because of a local one. But you know, will be a moment where some pattern would need this... I will start with something and later on, we can improve it while the API is still experimental.

Reply:
It's totally fine for me.